### DISCUSSION TITLE
Passing type into a recipe or cva instead of infer from them

### DISCUSSION THREAD

**Discussion by theMasix** [chakra-ui/panda#1513] [Ideas]
### Description

The only guide in the document tells us that there is a way to infer types from `cva` or `defineRecipe` and not pass types into them. I think there is a need to accept types from the outer scope and as a single source of truth for types of variants.

We already have the props (variants) type, Is there a way to use them as a validation type in recipes?

### Problem Statement/Justification

The problem is we can not use our pre-defined types as a source of truth for our PandaCSS cva or recipes and it may end up with type inconsistencies.

### Proposed Solution or API

`cva` or `defineRecipe` should accept type:

```tsx
const Button = cva<ButtonTypes>({})
// OR
const Button = defineRecipe<ButtonTypes>({})
```

**Comment by astahmer** [comment]
not sure I understand the purpose, can you show your use-case through a complete example ?

**Reply by theMasix** [reply]
Let's assume that we have the Button type as follows:

```tsx
type ButtonProps = {
  variant: 'outlined' | 'filled' | 'ghost';
  color: 'primary' | 'secondary' | 'danger' | 'warning'
}
```

then I want to pass this type to a `defineRecipe` method so the `variants` of a recipe could be defined "outside" of a recipe:

```tsx
const buttonRecipe = defineRecipe<ButtonProps>({
  variants: {
   variant: {
       // Accepts only 'outlined' | 'filled' | 'ghost';
    },
   color: {
      // Accepts only 'primary' | 'secondary' | 'danger' | 'warning'
   }
  }
})
```

This feature's main benefit is that a single truth source can be established and inconsistencies between button props and `defineRecipe` variants won't occur.

**Reply by astahmer** [reply]
you can make a wrapper like this:
```ts
const defineRecipeWithVariants = <T extends RecipeSelection<RecipeVariantRecord>>(config: RecipeConfig<{
  [K in keyof T]: T[K] extends string ? Record<T[K], SystemStyleObject> : never
}>) => config

```
<img width="856" alt="Screenshot 2023-10-15 at 14 41 22" src="https://github.com/chakra-ui/panda/assets/47224540/97a9a5d5-0c7e-49cd-9a6a-e237b488ccce">

**Reply by theMasix** [reply]
Great, I'll test the snippet. I'm wondering why there is no doc entry for this snippet in the official document.

**Reply by astahmer** [reply]
I didnt know about your use-case, that's why ! 😄  feel free to send a PR if you feel this is worth mentioning in the docs

**Reply by theMasix** [reply]
The type has an issue. boolean props can not be written in this way. Let's assume the `ButtonProps` has a `rounded: boolean` prop.

Then PandaCSS complains:

![Screenshot 2023-10-28 at 17 04 39](https://github.com/chakra-ui/panda/assets/22872117/17f0e8a6-e285-48f9-a2fe-70e0dbe761a1)

**Reply by theMasix** [reply]
I solved it by defining the `defineRecipeWithVariants` this way:

```ts
const defineRecipeWithVariants = <
  T extends RecipeSelection<RecipeVariantRecord>
>(
  config: RecipeConfig<{
    [K in keyof T]: T[K] extends string
      ? Record<T[K], SystemStyleObject>
      : T[K] extends boolean
      ? Record<'true' | 'false', SystemStyleObject>
      : never;
  }>
) => config;
```

Could I define better type?

**Reply by astahmer** [reply]
that looks fine to me 👍

**Reply by theMasix** [reply]
Could you please send me a type for `slotRecipe` as well?
